pkg/sessions/redis/redis_store.go

```diff
@@ -0,0 +1,271 @@
+	"strings"
+	"time"
+
+	"github.com/go-redis/redis"
+	"github.com/pusher/oauth2_proxy/cookie"
+	"github.com/pusher/oauth2_proxy/pkg/apis/options"
+	"github.com/pusher/oauth2_proxy/pkg/apis/sessions"
+	"github.com/pusher/oauth2_proxy/pkg/cookies"
+	"github.com/pusher/oauth2_proxy/pkg/sessions/utils"
+)
+
+// TicketData is a structure representing a used in server session storage
+type TicketData struct {
+	TicketID string
+	Secret   []byte
+}
+
+// SessionStore is an implementation of the sessions.SessionStore
+// interface that stores sessions in redis
+type SessionStore struct {
+	CookieCipher   *cookie.Cipher
+	CookieDomain   string
+	CookieExpire   time.Duration
+	CookieHTTPOnly bool
+	CookieName     string
+	CookiePath     string
+	CookieSecret   string
+	CookieSecure   bool
+	Client         *redis.Client
+}
+
+// NewRedisSessionStore initialises a new instance of the SessionStore from
+// the configuration given
+func NewRedisSessionStore(opts options.RedisStoreOptions, cookieOpts *options.CookieOptions) (sessions.SessionStore, error) {
+	opt, err := redis.ParseURL(opts.RedisConnectionURL)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse redis url: %s", err)
+	}
+
+	var cookieCipher *cookie.Cipher
+	if len(cookieOpts.CookieSecret) > 0 {
+		var err error
+		cookieCipher, err = cookie.NewCipher(utils.SecretBytes(cookieOpts.CookieSecret))
+		if err != nil {
+			return nil, fmt.Errorf("unable to create cookieCipher: %v", err)
+		}
+	}
+
+	client := redis.NewClient(opt)
+
+	rs := &SessionStore{
+		Client:         client,
+		CookieCipher:   cookieCipher,
+		CookieDomain:   cookieOpts.CookieDomain,
+		CookieExpire:   cookieOpts.CookieExpire,
+		CookieHTTPOnly: cookieOpts.CookieHTTPOnly,
+		CookieName:     cookieOpts.CookieName,
+		CookiePath:     cookieOpts.CookiePath,
+		CookieSecret:   cookieOpts.CookieSecret,
+		CookieSecure:   cookieOpts.CookieSecure,
+	}
+	return rs, nil
+
+}
+
+// Save takes a sessions.SessionState and stores the information from it
+// to redies, and adds a new ticket cookie on the HTTP response writer
+func (store *SessionStore) Save(rw http.ResponseWriter, req *http.Request, s *sessions.SessionState) error {
+	requestCookie, _ := req.Cookie(store.CookieName)
+	value, err := s.EncodeSessionState(store.CookieCipher)
+	if err != nil {
+		return err
+	}
+	ticketString, err := store.storeValue(value, s.ExpiresOn, requestCookie)
+	if err != nil {
+		return err
+	}
+
+	ticketCookie := cookies.MakeCookie(
+		req,
+		store.CookieName,
+		ticketString,
+		store.CookiePath,
+		store.CookieDomain,
+		store.CookieHTTPOnly,
+		store.CookieSecure,
+		store.CookieExpire,
+		time.Now(),
+	)
+
+	http.SetCookie(rw, ticketCookie)
+	return nil
+}
+
+// Load reads sessions.SessionState information from a ticket
+// cookie within the HTTP request object
+func (store *SessionStore) Load(req *http.Request) (*sessions.SessionState, error) {
+	requestCookie, _ := req.Cookie(store.CookieName)
+	// No cookie validation necessary
```

Comment:
I believe we sign the cookie in the cookie store implementation because we don't trust the storage - the user's browser, though I'm not entirely positive why we sign the session in any case, given that the interesting information is encrypted and always decrypted. The only place where that would be useful is if you wanted to just verify the signature and take for granted the plain text email/username in this scenario, but I don't see that in the code path. The other reason I could see why we do that is that the session issuance timestamp is in the clear, so maybe this needs to be moved into the session object.

In either case, in the redis storage the user has no access to those fields and cannot modify them. I can only think of two reasons why to sign it (and subsequently validate it)

* If we believe somebody can discover the session ID. In this implementation, the session ID (the ticket) is in fact a 256 bit random number, to which 128 of those bits are used to encrypt the session server side. We use `crypto/rand` too, so there's no issue there. 

* We want to reduce DoS attacks on the assumption that our connection to redis is susceptible to them. This is probably best handled outside by other infrastructure (web application firewalls, fail2ban, etc...) and it's not so likely to be a real issue for redis, which can easily handle 10k+ queries a second.

* If we want to timestamp when _the cookie_ was issued to the user (but not when the session was created/modified). I think this is of little value in the redis store, because redis will expire the session if it's out of date. We also have an `ExpiresOn` field in `SessionState`, though what may be helpful is an additional `Timestamp` or an `Issued` `time.Time` field to the `SessionState` struct and rely on that for validation for  session stores. In this scenario, we would `validateSession` instead of `validateCookie`.

Beyond that, I suppose you might want to do that to prevent a timing attack on old (pre-2008) or exotic processors (Go promises to use constant time encryption algorithms on AES-NI enabled processors - https://golang.org/pkg/crypto/aes/), but the mac comparison would also need to be constant time.

Reply:
That's a pretty convincing argument, I'm happy to leave without signatures 👍 Thanks